plugins/debuginfo-install.py

```diff
@@ -74,59 +75,125 @@ def configure(self):
         self.base.repos.enable_debug_repos()
 
     def run(self):
-        self.packages = self.base.sack.query()
-        self.packages_available = self.packages.available()
         errors_spec = []
 
+        debuginfo_suffix_len = len(Package.DEBUGINFO_SUFFIX)
+        debugsource_suffix_len = len(Package.DEBUGSOURCE_SUFFIX)
+
         for pkgspec in self.opts.package:
-            package_query = sorted(dnf.subject.Subject(pkgspec).get_best_query(
-                self.cli.base.sack).filter(arch__neq='src'), reverse=True)
-            if not package_query:
-                msg = _('No match for argument: %s')
-                logger.info(msg, self.base.output.term.bold(pkgspec))
+            solution = dnf.subject.Subject(pkgspec).get_best_solution(self.base.sack)
+
+            query = solution["query"].filterm(arch__neq='src')
+
+            package_dict = query.available()._name_dict()
+            # installed versions of packages have priority, replace / add them to the dict
+            package_dict.update(query.installed()._name_dict())
+
+            if not package_dict:
+                logger.info(_('No match for argument: %s'), self.base.output.term.bold(pkgspec))
                 errors_spec.append(pkgspec)
-            for pkg in package_query:
-                self._di_install(pkg)
+                continue
 
-        if not self.dbgdone:
-            logger.info(_("No debuginfo packages available to install"))
+            # Remove debuginfo packages if their base packages are in the query.
+            # They can get there through globs and they break the installation
+            # of debug packages with the same version as the installed base
+            # packages. If the base package of a debuginfo package is not in
+            # the query, the user specified a debug package on the command
+            # line. We don't want to ignore those, so we will install them.
+            # But, in this case the version will not be matched to the
+            # installed version of the base package, as that would require
+            # another query and is further complicated if the user specifies a
+            # version themselves etc.
+            for name in list(package_dict.keys()):
+                if name.endswith(Package.DEBUGINFO_SUFFIX):
+                    if name[:-debuginfo_suffix_len] in package_dict:
```

Comment:
Please add condition for testing arch. This could be very tricky because foo-debuginfo could be in dist with a single architecture and foo could have also a single or multiple arch packages in dict.

Reply:
Not sure I understand you, but if you mean I should test if the `-debuginfo` package in the dict has the same architecture as the base package I'm looking up here, I don't think it's needed.

I don't think we really care about the architecture now, we are just detecting a case when both `-debuginfo` and base package matched a single input pkgspec. And if it did, we want to remove the `-debuginfo` package, becase it is superfluous here and breaks the correct handling of installing the debuginfo for the base package later, which will take care of the architectures too...

It's a bit complicated but I think there is no problem here, in case you still think it is, can you think of a test case I could add for it?

Reply:
Probably you all right here.